src/main/java/com/javax0/jscc/Compiler.java

```diff
@@ -26,89 +28,99 @@
+     * @throws Exception
+     *             exceptions loading the class are not caught by the method
+     */
+    public Class<?> compile(String sourceCode, String canonicalClassName) throws Exception {
+	JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+	List<JavaSourceFromString> sources = new LinkedList<>();
+	String className = calculateSimpleClassName(canonicalClassName);
+	sources.add(new JavaSourceFromString(className, sourceCode));
 
-	/**
-	 * Compile the Java code provided as a string and if the compilation was
-	 * successful then load the class.
-	 * 
-	 * @param sourceCode
-	 * @param canonicalClassName
-	 *            the fully qualified name of the class
-	 * @return the loaded class or null if the compilation was not successful
-	 * @throws Exception
-	 *             exceptions loading the class are not caught by the method
-	 */
-	public Class<?> compile(String sourceCode, String canonicalClassName)
-			throws Exception {
-		JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
-		List<JavaSourceFromString> sources = new LinkedList<>();
-		String className = calculateSimpleClassName(canonicalClassName);
-		sources.add(new JavaSourceFromString(className, sourceCode));
-
-		StringWriter sw = new StringWriter();
-		MemoryJavaFileManager fm = new MemoryJavaFileManager(
-				compiler.getStandardFileManager(null, null, null));
-		JavaCompiler.CompilationTask task = compiler.getTask(sw, fm, null,
-				null, null, sources);
-
-		Boolean compilationWasSuccessful = task.call();
-		if (compilationWasSuccessful) {
-			ByteClassLoader byteClassLoader = new ByteClassLoader(new URL[0],
-					classLoader, classesByteArraysMap(fm));
-
-			Class<?> klass = byteClassLoader.loadClass(canonicalClassName);
-			byteClassLoader.close();
-			return klass;
-		} else {
-			compilerErrorOutput = sw.toString();
-			return null;
-		}
+	StringWriter sw = null;
+	if (diagnostics == null) {
+	    sw = new StringWriter();
 	}
+	MemoryJavaFileManager fm = new MemoryJavaFileManager(compiler.getStandardFileManager(diagnostics, null, null));
+	JavaCompiler.CompilationTask task = compiler.getTask(sw, fm, diagnostics, optionList, null, sources);
 
-	/**
-	 * Get the map of class name / class byte array from the file manager.
-	 * 
-	 * @param fileManager
-	 * @return
-	 */
-	private Map<String, byte[]> classesByteArraysMap(
-			MemoryJavaFileManager fileManager) {
-		Map<String, byte[]> map = new HashMap<>();
-		for (String name : fileManager.getClassFileObjectsMap().keySet()) {
-			map.put(name, fileManager.getClassFileObjectsMap().get(name)
-					.getByteArray());
-		}
-		return map;
+	Boolean compilationWasSuccessful = task.call();
+	if (compilationWasSuccessful) {
+	    ByteClassLoader byteClassLoader = new ByteClassLoader(new URL[0], classLoader, classesByteArraysMap(fm));
+
+	    Class<?> klass = byteClassLoader.loadClass(canonicalClassName);
+	    byteClassLoader.close();
+	    return klass;
+	}
+	if (sw != null) {
```

Comment:
Deleting `else` online 88 makes the code less elegant and less readable. Usually there has to be only one single return statement in a method. In some cases there can be more, like

1. a success return
1. an error return

In this case the structure is

```method(){
   ..... something to calculate

  if( success condition ){
    return success
    }else{
    return error
   }
}```